### ISSUE TITLE
Always-on VPN and Airplane mode

### ISSUE THREAD

**Issue by ignoramous** [celzero/rethink-app#1157]
A [user writes](https://mastodon.social/@kuketzblog@social.tchncs.de/111512511888998159),

> What I noticed with RethinkDNS: If the "Continuously active VPN" option is active, apps such as K-9 Mail constantly try to establish a push connection - even when the device is in flight mode. The apps probably think there is still a connection to the Internet via VPN, although there definitely isn't one. This calls for a fix from RethinkDNS.

**Comment by ignoramous** [comment]
Thunderbird: https://github.com/thunderbird/thunderbird-android/issues/7408

**Comment by ignoramous** [comment]
On Android 8 and below: In airplane mode (or when no network has internet), `VpnService.setUnderlyingNetworks(emptyArray())` ([docs](http://developer.android.com/reference/android/net/VpnService.Builder.html#setUnderlyingNetworks(android.net.Network%5B%5D))) signals to apps (ex: Chrome) that there is absolutely no way to connect.

On Android 9 and above: Neither `VpnService.setUnderlyingNetworks(emptyArray())` nor setting just the default gateway (but not the DNS) works.

**Comment by ItsIgnacioPortal** [comment]
For anyone interested, I'm willing to engineer fix for this for USD$30 (Not an android-system fix, but a RethinkDNS-specific fix).

**Comment by ignoramous** [comment]
> For anyone interested, I'm willing to engineer fix for this for USD$30 (Not an android-system fix, but a RethinkDNS-specific fix).

Thanks. Please note that it is no guarantee Rethink will merge that fix.

**Comment by ItsIgnacioPortal** [comment]
> > For anyone interested, I'm willing to engineer fix for this for USD$30 (Not an android-system fix, but a RethinkDNS-specific fix).
> 
> Thanks. Please note that it is no guarantee Rethink will merge that fix.

The fix I have in mind won't modify any RethinkDNS code.

**Comment by ignoramous** [comment]
Interestingly, Daniel Micay posits `VpnService.setUnderylingNetworks` should work: https://github.com/GrapheneOS/os-issue-tracker/issues/3921 May be we should give it another try.

**Comment by Grishnackh** [comment]
I would appreciate when you give this another try.

Obtainium implemented a fix that's looking more like a workaround with that additional information. Other apps, i.e. the GrapheneOS updater is still affected and gives error messages when the phone is in airplane mode + connected to RethinkDNS VPN.

**Comment by ignoramous** [comment]
Implemented in `v055p`+.

Turn ON *Configure -> Network -> Stall on network loss*, after much wrangling with `VpnService.setUnderlyingNetworks`.

---

> allowing fast switching between networks when the network changes

`VpnService.setUnderlyingNetworks` isn't necessary for "fast switching", but only necessary if you'd want to expose the capabilities of all underlying networks as the VPN's (that is, declare to apps listening on Network events that the VPN supports both WiFi and Mobile at the same time, though it is unclear how apps can ask the VPN to bind to those networks *without* VPN allowing it via `VpnService.allowBypass`, which will NOT work in VPN *Lockdown* mode). 

> You have immediate switching between Wi-Fi and cellular networks and vice versa. This is very convenient, especially if you use Tor, as it does not like interruptions

For uninterrupted connectivity (on network changes or when recreating the VPN tunnel to get a new `fd`, for example), Android supports "seamless handover" where the previous tun `fd` can be read until EOF alongside the newly created tun `fd`. In our experiments, this works, but not in VPN *Lockdown* mode ("Block connections without VPN" turned ON).

Leaving `VpnService.setUnderlyingNetworks` to `null`, Android will set the actual underlying network for the VPN tunnel to `ConnectivityManager.activeNetwork` in all circumstances. Which, along with "seamless handover", might just be enough.

---

> feature is very fragile and has to be used a very specific way in order to work properly. 

Yes, which is why we've disabled it by default. It seems to work differently across different OEMs and ROMs, but Rethink's bar with the latest release was to make it work on Graphene, if no where else.

> Not sure on the exact details of that. RehinkDNS tried to adopt it recently and had a lot of issues.

The issues were resolved by:
- Calling `VpnService.setUnderlyingNetwork` on every capability change of whatever *Network* objects were previously set as underlying network.
- Not attempting "seamless handover" when in VPN *Lockdown* mode.
- When no validated (internet-capable) networks are available, setting `VpnService.setUnderlyingNetwork(null)` as the underlying network instead of an empty array, `[]`, when in VPN *Lockdown* mode (but this is changeable by the end user, if it works on their Android version / ROM).
- Never setting an unvalidated network as the underlying network when in VPN *Lockdown* mode (but this may break LAN connectivity over networks that won't have *default* routes to the Internet).
- Binding to one of the networks set as the underlying network, explicitly, for egress sockets created by Rethink on behalf of apps (again, only in VPN *Lockdown* mode).
- (I may be forgetting other stuff but the code is up, and it isn't the prettiest from handling a myriad of configurations, as you can imagine).

I'd only recommend `VpnService.setUnderlyingNetworks` to my worst enemy (:

**Comment by ignoramous** [comment]
Also:
- https://github.com/GrapheneOS/os-issue-tracker/issues/3921#issuecomment-3172941295

**Comment by alexanderadam** [comment]
Hello,

following up here, since I ran into this issue [from a different angle](https://codeberg.org/iNPUTmice/Conversations/issues/792#issue-3445424).

Maybe there's still some room for improvement, even with the "Stall on network loss" setting.

With the default config ([`stallOnNoNetwork` defaulting to `false`](https://github.com/celzero/rethink-app/blob/04369d86fad4bd76d3fc8568c3226177f4efba60/app/src/main/java/com/celzero/bravedns/service/PersistentState.kt#L398)), [`setUnderlyingNetwork(null)` is called](https://developer.android.com/reference/android/net/VpnService#setUnderlyingNetworks(android.net.Network%5B%5D)) ([coming from `getUnderlays`](https://github.com/celzero/rethink-app/blob/04369d86fad4bd76d3fc8568c3226177f4efba60/app/src/main/java/com/celzero/bravedns/service/BraveVPNService.kt#L2989-L2997)). So basically telling apps to use whatever netwrok is around and having apps doing their connection attempts despite us knowing that it won't work anyway.
Which is not great for battery runtime .

It seems that [Netguard](https://github.com/M66B/NetGuard) is doing this much more elegant (they also had the `setUnderlyingNetworks` but [it's disabled now](https://github.com/M66B/NetGuard/blob/7308869411ff87649bf3a46a9c7c08f1e5353801/app/src/main/java/eu/faircode/netguard/ServiceSinkhole.java#L1246)):
On netwrok changes (i.e. [`NetworkCallback.onLost()`](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onLost(android.net.Network))) it [calls `reload` and within `getAllowedRules`](https://github.com/M66B/NetGuard/blob/master/app/src/main/java/eu/faircode/netguard/ServiceSinkhole.java#L565) which [excludes the VPN](https://github.com/M66B/NetGuard/blob/7308869411ff87649bf3a46a9c7c08f1e5353801/app/src/main/java/eu/faircode/netguard/Util.java#L173) explicitly in the check.

Thus no apps are allowed to connect anymore because this already drops the packages and network layer and stops retrying sooner.

Furthermore, right now Rethink only checks for `NetworkCallback.onLost()` and additionally listening to [`ACTION_AIRPLANE_MODE_CHANGED`](https://developer.android.com/reference/android/content/Intent#ACTION_AIRPLANE_MODE_CHANGED) might catch the broadcast earlier without having to wait for for the `onLast`.

**PS:** Maybe it's just me, but the names [`stallOnNoNetwork` and `fail_open_on_no_network` are confusing](https://github.com/celzero/rethink-app/blob/04369d86fad4bd76d3fc8568c3226177f4efba60/app/src/main/java/com/celzero/bravedns/service/PersistentState.kt#L398). It kinda sounds like they're the opposite, no?

**Comment by ignoramous** [comment]
> Maybe it's just me, but the names stallOnNoNetwork and fail_open_on_no_network are confusing. It kinda sounds like they're the opposite, no?

Yeah, the key for sharedpreference was named before we changed to *Stall on network loss* from previously named setting *Do not stall on network loss*. We kept it.

> Furthermore, right now Rethink only checks for NetworkCallback.onLost() and additionally listening to ACTION_AIRPLANE_MODE_CHANGED might catch the broadcast earlier 

Airplane mode doesn't mean network loss. For instance, wifi can still be connected.

> On netwrok changes (i.e. NetworkCallback.onLost()) it calls reload and within getAllowedRules which excludes the VPN explicitly in the check

Not sure what this means or does? I looked at the code but it isn't clear what you mean by "excludes VPN explicitly"? By default, all VPN apps are "excluded" automatically on Android 5+ from the VPN's own tunnel. Rethink targets Android 6+

> default config (stallOnNoNetwork defaulting to false), setUnderlyingNetwork(null) is called (coming from getUnderlays). So basically telling apps to use whatever netwrok is around

As I understand it, `setUnderlyingNetwork` sets underlay for the VPN app, not for other apps. Though, you would be right that these underlay networks affect how the apps may react to `ConnectivityManager` sent events.

**Comment by alexanderadam** [comment]
Woah! You are super-fast with responding! 🙏 

> > Furthermore, right now Rethink only checks for NetworkCallback.onLost() and additionally listening to ACTION_AIRPLANE_MODE_CHANGED might catch the broadcast earlier
> 
> Airplane mode doesn't mean network loss. For instance, wifi can still be connected.


yes, fair point. I was sloppy equating the two. 😉 

> > On netwrok changes (i.e. NetworkCallback.onLost()) it calls reload and within getAllowedRules which excludes the VPN explicitly in the check
> 
> Not sure what this means or does? I looked at the code but it isn't clear what you mean by "excludes VPN explicitly"? By default, all VPN apps are "excluded" automatically on Android 5+ from the VPN's own tunnel. Rethink targets Android 6+

Sorry, I was unclear. The only point that I wanted to make here, is that they just stop forwarding traffic. 😉 

Would it be possible to combine the events of `ACTION_AIRPLANE_MODE_CHANGED`, `onLost`, `onAvailable` and then update the state with the available network checks within?

Like, maybe [`getUnderlays`](https://github.com/celzero/rethink-app/blob/04369d86fad4bd76d3fc8568c3226177f4efba60/app/src/main/java/com/celzero/bravedns/service/BraveVPNService.kt#L2987) and [`determineRoutes`](https://github.com/celzero/rethink-app/blob/04369d86fad4bd76d3fc8568c3226177f4efba60/app/src/main/java/com/celzero/bravedns/service/BraveVPNService.kt#L3553) could be changed to


```
val airplaneAndNoNetwork = isAirplaneModeOn && !hasUnderlyingNetwork
val failOpen = !persistentState.stallOnNoNetwork && !airplaneAndNoNetwork
```

The `isAirplaneModeOn` would be set on `ACTION_AIRPLANE_MODE_CHANGED`.

I mean, the other cases might still cause weird behaviour but if we got the event change that Airplane mode was activated _and_ we have no networks, then the case should be kinda clear, no?

**Comment by ignoramous** [comment]
> The only point that I wanted to make here, is that they just stop forwarding traffic. 

Believe `VpnService.setUnderlyingNetworks([])` (setting underlays to empty array) achieves the same result, which Rethink has guarded behind *Configure -> Network -> Stall on network loss* setting, because the behaviour of doing so is wildly different across Android versions and OEMs (Pixels on Android 12, for example, work fine; OnePlus on Android 12, didn't... as in OnePlus would never recover and lose connectivity permanently regardless of network connectivity changes and the only way out was to turn off the VPN app).

> the other cases might still cause weird behaviour but if we got the event change that Airplane mode was activated and we have no networks

Are you making a case of setting underlying network to an empty array even when *Configure -> Network -> Stall on network loss* is turned OFF? If so, we had a very bad experience with this from `v055p` through `v055q` (see above and [this comment](https://github.com/celzero/rethink-app/issues/1157#issuecomment-3687483512)).

**Comment by alexanderadam** [comment]
> > the other cases might still cause weird behaviour but if we got the event change that Airplane mode was activated and we have no networks
> 
> Are you making a case of setting underlying network to an empty array even when _Configure -> Network -> Stall on network loss_ is turned OFF? If so, we had a very bad experience with this from `v055p` through `v055q` (see above and [this comment](https://github.com/celzero/rethink-app/issues/1157#issuecomment-3687483512)).

Yes but only for the narrow case when airplane mode is on and there are no networks available.
Do I understand that correctly OnePlus never recovers even if a later call tries to override this again?

Or Rethink would send things like `RST`, `SERVFAIL` etc directly. But this would be complicated for sure.

**Comment by ignoramous** [comment]
>  Do I understand that correctly OnePlus never recovers even if a later call tries to override this again?

Yeah, the device we had didn't (thankfully may I add, as that's how we even figured out that some ROMs / OEMs had broken implementation). And there were countless other reports from F-Droid users who updated to `v055p` & `v055q`.
